agent/src/capability/configstore.cpp

```diff
@@ -0,0 +1,237 @@
+        return ReturnCode::FAILURE;
+    }
+    if (isDirectoryEmpty(dirPath)) {
+        log_warning() << "Path to configuration files is empty: " << dirPath;
+        return ReturnCode::FAILURE;
+    }
+
+    std::vector<std::string> files;
+    files = fileList(dirPath);
+
+    if (files.empty()) {
+        log_info() << "No configuration files found: " << dirPath;
+        return ReturnCode::SUCCESS;
+    }
+
+    for (unsigned int i = 0; i != files.size(); i++) {
+        std::string filePath = dirPath + files[i];
+
+        if (readCapsFromFile(filePath) != ReturnCode::SUCCESS) {
+            log_warning() << "Could not parse a file in directory: " << filePath;
+        }
+    }
+
+    return ReturnCode::SUCCESS;
+}
+
+
+GatewayConfiguration ConfigStore::readConfigs(const std::string capID)
+{
+    return m_configStore[capID];
+}
+
+
+ReturnCode ConfigStore::parseCapabilities(json_t *capabilities)
+{
+    std::map<std::string, GatewayConfiguration> caps;
+
+    size_t i;
+    json_t *capability;
+
+    json_array_foreach(capabilities,i,capability) {
+        if (!json_is_object(capability)) {
+            log_error() << "Capability is not a json object";
+            return ReturnCode::FAILURE;
+        }
+
+        std::string capName;
+        if (!JSONParser::read(capability,"name",capName)) {
+            log_error() << "Could not read Capability name";
+            return ReturnCode::FAILURE;
+        }
+
+        json_t *gateways = json_object_get(capability, "gateways");
+        if (gateways == nullptr) {
+            log_error() << "Could not read Gateways";
+            return ReturnCode::FAILURE;
+        }
+        if (!json_is_array(gateways)) {
+            log_error() << "Gateways is not an array";
+            return ReturnCode::FAILURE;
+        }
+
+        return parseGateways(capName,gateways);
+    }
+    return ReturnCode::SUCCESS;
+}
+
+ReturnCode ConfigStore::parseGateways(std::string capName, json_t *gateways)
+{
+    size_t j;
+    json_t *gateway;
+
+    GatewayConfiguration gwConfs;
+
+    json_array_foreach(gateways, j, gateway) {
+        if (!json_is_object(gateway)) {
+            log_error() << "Gateway is not a json object";
+            return ReturnCode::FAILURE;
+        }
+
+        std::string gwID;
+        if (!JSONParser::read(gateway,"id",gwID)) {
+            log_error() << "Could not read Gateway ID";
+            return ReturnCode::FAILURE;
+        }
+
+        json_t *confs = json_object_get(gateway,"config");
+        if (confs == nullptr) {
+            log_error() << "Could not read Gateway Config";
+            return ReturnCode::FAILURE;
+        }
+        if (!json_is_array(confs)) {
+            log_error() << "Gateway Config is not an array";
+            return ReturnCode::FAILURE;
+        }
+        gwConfs[gwID] = confs;
+
+    }
+
+    GatewayConfiguration oldGwConfs = m_configStore[capName];
```

Comment:
Isn't the point here to **not fail** even if there are no old configs. The [] operator on std::map constructs an object of the value type if there is none available. I think it can be kept as is.

If there is a nice way to make it explicit if there are any old configs or not without making it too complicated, that could also be nice of course.